portal/flatpak-portal.c

```diff
@@ -999,7 +1029,28 @@ handle_spawn (PortalFlatpak         *object,
       env = g_strdupv (empty);
     }
   else
-    env = g_get_environ ();
+    {
+      GStrv run_environ;
+
+      if (testing)
+        {
+          static char *mock_run_environ[] = { "FOO=bar", NULL };
+          run_environ = mock_run_environ;
+        }
+      else
+        {
+          run_environ = flatpak_instance_get_run_environ (instance);
+          if (!run_environ)
+            {
+              g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR,
+                                                     G_DBUS_ERROR_INVALID_ARGS,
+                                                     "Could not find run environment of caller instance");
+              return G_DBUS_METHOD_INVOCATION_HANDLED;
```

Comment:
This can legitimately happen if you start an app with an older Flatpak, then restart the portal with a newer Flatpak (or while testing a newer Flatpak), then the running app asks for a subsandbox via the portal. I think we should `g_info()` and fall back to the portal's `g_get_environ ()` here. Perhaps like this?

```
static const char * const mock_run_environ[] = { "FOO=bar", NULL };
const char * const *run_environ;

if (testing)
    run_environ = mock_run_environ;
else
    run_environ = flatpak_instance_get_run_environ (instance);

if (run_environ != NULL)
  {
    env = g_strdupv ((const char * const *) run_environ);
  }
else
  {
    g_info ("Could not find `flatpak run` environment from caller instance, falling back to portal's environment");
    env = g_get_environ ();
  }

Reply:
I don't mind doing this, but the reason I decided not to do this originally is because I thought the case of restarting Flatpak services like the portal without also restarting apps after updating Flatpak is unlikely/uninteresting/unsupported.

For example, if you restart `flatpak-session-helper` you already lose some functionality in already-running apps in the form of the p11-kit socket becoming nonfunctional. It's easy to see in every Chromium-based app that every certificate is suddenly considered untrusted until the app is restarted.

The benefit of not doing this and instead treating this case as a fatal error is that this functionality could never be silently broken in the case of a bug or otherwise.

With that in mind, I'll leave the decision to you.

Reply:
The session helper and the portal are a bit different, and coping with existing apps still running is easy to achieve here (if anything I think it's *fewer* lines of code to do as I suggested), so I have a weak preference for making it resilient here.

Reply:
I guess I'm trying to say that I consider the the lack of fallback in this case to be a feature, not an omission or an oversight.

We would never allow subsandboxes to be established based on the portal's environment, regardless of which obnormal flow we end up in. Eliminating the option to enter such an edge case first place avoids the potential for any unpredictable behavior surrounding it. Questions like "maybe that issue happened because we ended up inheriting the portal's environment instead of flatpak-run's" would never have to come to anyone's mind.

I'll be honest, I would prefer not to add a fallback for this case if your preference to the contrary is only weak, and if you consider it acceptable. I will of course accept any decision, however.

I can follow your suggestion to simplify the code and save a few lines of code regardless.

Reply:
I haven't had much time to look into this, so for now I only simplified the code according to your suggestion without changing its behavior.

Feel free to resolve this thread if you agree with my design choice of forbidding a fallback to the old behavior. If you disagree with this design, please mention it in a comment and I'll implement the fallback.

Reply:
I think yes we do need to handle restarting the portal with a Flatpak app already running.

As well as being robust against restarting services during debugging, here is a scenario in which it matters:

1. Launch Steam as a Flatpak app, using Flatpak 1.14. Its environment isn't saved, because it's an old version of Flatpak. Also, `flatpak-portal` is not yet running, because nothing has asked for it yet.
2. Play a legacy native Linux game that does not use Steam Linux Runtime for a while.
3. Meanwhile, OS services upgrade the `flatpak` package to 1.15.x (including the change you propose here). In a development OS that doesn't do atomic upgrades, this doesn't necessarily need to imply a reboot.
4. Start a Steam Linux Runtime or Proton game.
5. `flatpak-portal` is autostarted by D-Bus activation. It's version 1.15.x now!
6. Desired result: the SLR or Proton game still starts successfully, even though there is no environment for it to load. Result with this PR as currently implemented: the SLR or Proton game doesn't start.

For visibility, I think it would be OK to log at `g_warning()` level here rather than `g_info()`, so that the message will be logged to the Journal by default. I also think it would be OK to only tolerate `ENOENT`, and fail hard (as you have advocated) on any other error.

Untested implementation:

```c
    {
      static const char * const mock_run_environ[] = { "FOO=bar", NULL };
      g_auto(GStrv) run_environ = NULL;

      if (testing)
        env = g_strdupv ((GStrv) mock_run_environ);
      else
        env = flatpak_instance_get_run_environ (instance, error);

      if (env == NULL)
        {
          if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
            {
              g_warning ("Environment of `flatpak run` was not saved, falling back to current environment");
              env = g_get_environ ();
            }
          else
            {
              g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR,
                                                     G_DBUS_ERROR_INVALID_ARGS,
                                                     "%s", error->message);
              return G_DBUS_METHOD_INVOCATION_HANDLED;
            }
        }
    }
```

(We could test this by starting a Flatpak app with the proposed version, and then either deleting its `run-environ` or chmod'ing it 000 before trying to start a subsandbox.)

Reply:
Done and tested per your suggestions.

As a side-note, it seems the correct check is `g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT)`.